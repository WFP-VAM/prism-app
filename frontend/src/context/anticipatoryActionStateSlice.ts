import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { LayerDefinitions } from 'config/utils';
import { AnticipatoryActionLayerProps } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from './store';

export const allWindowsKey = 'All';

// na/ny are not actually found in CSV, but defined not to cause confusion when calling the functions
export const AAcategory = ['na', 'ny', 'Leve', 'Moderado', 'Severo'] as const;
export type AACategoryType = typeof AAcategory[number];

export const AAPhase = ['na', 'ny', 'Ready', 'Set'] as const;
export type AAPhaseType = typeof AAPhase[number];

export interface AnticipatoryActionData {
  category: AACategoryType;
  district: string;
  index: string;
  month: string;
  phase: AAPhaseType;
  probability: string;
  trigger: string;
  triggerNB: string;
  triggerType: string;
  type: string;
  windows: string;
  yearOfIssue: string;
  date: string;
}

export const AAlayerKey = 'anticipatory_action';

const AACSVKeys: [string, string][] = [
  ['Category', 'category'],
  ['District', 'district'],
  ['Index', 'index'],
  ['Month', 'month'],
  ['Phase', 'phase'],
  ['Probability', 'probability'],
  ['Trigger', 'trigger'],
  ['Trigger_nb', 'triggerNB'],
  ['Trigger_type', 'triggerType'],
  ['Type', 'type'],
  ['Windows', 'windows'],
  ['Year_of_issue', 'yearOfIssue'],
];

function transform(data: any[], keys: [string, string][]) {
  const parsed = data
    .filter(x => !!x[keys[0][0]]) // filter empty rows
    .map(obj => {
      const entries = keys.map(k => [k[1], obj[k[0]]]);
      const month = obj.Month.padStart(2, '0');
      const year = obj.Year_of_issue.split('-')[0];

      // TODO: update this once CSV date format is revised
      const actualYear = Number(month) < 5 ? String(Number(year) + 1) : year;
      const date = `${actualYear}-${month}-01`;
      return Object.fromEntries([...entries, ['date', date]]);
    }) as AnticipatoryActionData[];

  const windows = [...new Set(parsed.map(x => x.windows))];
  // eslint-disable-next-line fp/no-mutating-methods
  const availableDates = [
    ...new Set(parsed.map(x => new Date(x.date).getTime())),
  ].sort();

  const groupedByDistrict = new Map<string, AnticipatoryActionData[]>();

  parsed.forEach(x => {
    const { district } = x;
    const rows = groupedByDistrict.get(district);
    groupedByDistrict.set(district, rows ? [...rows, x] : [x]);
  });

  const sortFn = (a: AnticipatoryActionData, b: AnticipatoryActionData) => {
    const aDate = new Date(a.date).valueOf();
    const bDate = new Date(b.date).valueOf();
    if (aDate > bDate) {
      return -1;
    }
    if (bDate > aDate) {
      return 1;
    }
    const aCatIndex = AAcategory.findIndex(x => x === a.category);
    const bCatIndex = AAcategory.findIndex(x => x === b.category);
    if (aCatIndex > bCatIndex) {
      return -1;
    }
    if (bCatIndex > aCatIndex) {
      return 1;
    }
    return 0;
  };

  const result = Object.fromEntries(
    // eslint-disable-next-line fp/no-mutating-methods
    Array.from(groupedByDistrict.entries()).map(x => [x[0], x[1].sort(sortFn)]),
  );

  const monitoredDistricts = Object.keys(result);

  return { data: result, windows, availableDates, monitoredDistricts };
}

type AnticipatoryActionState = {
  data: {
    [k: string]: AnticipatoryActionData[];
  };
  availableDates?: number[];
  monitoredDistricts: string[];
  windows: string[];
  selectedWindow: string;
  categoryFilters: Record<AACategoryType, boolean>;
  loading: boolean;
  error: string | null;
};

const initialState: AnticipatoryActionState = {
  data: {},
  windows: [],
  availableDates: undefined,
  monitoredDistricts: [],
  selectedWindow: allWindowsKey,
  categoryFilters: {
    Severo: true,
    Moderado: true,
    Leve: true,
    na: false,
    ny: false,
  },
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    data: {
      [k: string]: AnticipatoryActionData[];
    };
    windows: string[];
    availableDates: number[];
    monitoredDistricts: string[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const layer = LayerDefinitions[AAlayerKey] as AnticipatoryActionLayerProps;

  const url = layer.baseUrl;
  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(transform(results.data, AACSVKeys)),
      error: error => reject(error),
    }),
  );
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setSelectedWindow: (state, { payload }: PayloadAction<string>) => ({
      ...state,
      selectedWindow: payload,
    }),
    setCategoryFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<Record<'Leve' | 'Moderado' | 'Severo', boolean>>
      >,
    ) => ({
      ...state,
      categoryFilters: { ...state.categoryFilters, ...payload },
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.data,
      windows: payload.windows,
      availableDates: payload.availableDates,
      monitoredDistricts: payload.monitoredDistricts,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AnticipatoryActionDataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AnticipatoryActionWindowsSelector = (state: RootState) =>
  state.anticipatoryActionState.windows;

export const AnticipatoryActionAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AASelectedWindowSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedWindow;

export const AACategoryFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.categoryFilters;

// export actions
export const {
  setSelectedWindow,
  setCategoryFilters,
} = anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
