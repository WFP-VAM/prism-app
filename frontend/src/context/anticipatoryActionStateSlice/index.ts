import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { DateItem } from 'config/types';
import { appConfig } from 'config';
import { AAWindowKeys } from 'config/utils';
import type { CreateAsyncThunkTypes, RootState } from '../store';
import {
  AACategoryType,
  AAView,
  AnticipatoryActionData,
  AnticipatoryActionState,
  Vulnerability,
  allWindowsKey,
} from './types';
import {
  calculateMapRenderedDistricts,
  emptyWindows,
  parseAndTransformAA,
} from './utils';

const initialState: AnticipatoryActionState = {
  data: emptyWindows,
  availableDates: undefined,
  monitoredDistricts: [],
  filters: {
    selectedDate: undefined,
    selectedWindow: allWindowsKey,
    selectedIndex: '',
    categories: {
      Severe: true,
      Moderate: true,
      Mild: true,
      Normal: true,
      na: true,
      ny: true,
    },
  },
  markers: [],
  selectedDistrict: '',
  renderedDistricts: emptyWindows,
  windowRanges: { 'Window 1': undefined, 'Window 2': undefined },
  view: AAView.Home,
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    windowData: {
      data: AnticipatoryActionData;
      availableDates: DateItem[];
      windowKey: (typeof AAWindowKeys)[number];
      range: { start: string; end: string };
    }[];
    monitoredDistricts: { name: string; vulnerability: Vulnerability }[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const url = appConfig.anticipatoryActionUrl;

  return new Promise<any>((resolve, reject) => {
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(parseAndTransformAA(results.data)),
      error: error => reject(error),
    });
  });
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setAAFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<{
          selectedDate: string | undefined;
          selectedWindow: (typeof AAWindowKeys)[number] | typeof allWindowsKey;
          selectedIndex: string;
          categories: Partial<Record<AACategoryType, boolean>>;
        }>
      >,
    ) => {
      const { categories, ...rest } = payload;
      const newCategories =
        categories !== undefined
          ? { ...state.filters.categories, ...categories }
          : state.filters.categories;
      const newFilters = {
        ...state.filters,
        ...rest,
        categories: newCategories,
      };
      return {
        ...state,
        filters: newFilters,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: newFilters,
          data: state.data,
          windowRanges: state.windowRanges,
        }),
      };
    },
    setAASelectedDistrict: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['selectedDistrict']>,
    ) => ({
      ...state,
      selectedDistrict: payload,
    }),
    setAAMarkers: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['markers']>,
    ) => ({
      ...state,
      markers: payload,
    }),
    setAAView: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['view']>,
    ) => ({
      ...state,
      view: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => {
      const newData = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.data]),
      ) as Record<(typeof AAWindowKeys)[number], AnticipatoryActionData>;
      const newRanges = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.range]),
      ) as Record<
        (typeof AAWindowKeys)[number],
        { start: string; end: string }
      >;
      return {
        ...state,
        loading: false,
        data: newData,
        availableDates: Object.fromEntries(
          payload.windowData.map(x => [x.windowKey, x.availableDates]),
        ) as Record<(typeof AAWindowKeys)[number], DateItem[]>,
        windowRanges: newRanges,
        monitoredDistricts: payload.monitoredDistricts,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: state.filters,
          data: newData,
          windowRanges: newRanges,
        }),
      };
    });

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error: _error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AADataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AAAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AAMonitoredDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.monitoredDistricts;

export const AAFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.filters;

export const AARenderedDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.renderedDistricts;

export const AASelectedDistrictSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedDistrict;

export const AAMarkersSelector = (state: RootState) =>
  state.anticipatoryActionState.markers;

export const AAViewSelector = (state: RootState) =>
  state.anticipatoryActionState.view;

export const AAWindowRangesSelector = (state: RootState) =>
  state.anticipatoryActionState.windowRanges;

// export actions
export const { setAAFilters, setAASelectedDistrict, setAAMarkers, setAAView } =
  anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
