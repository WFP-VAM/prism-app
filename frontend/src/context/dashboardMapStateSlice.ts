import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Map as MaplibreMap, LngLatBounds } from 'maplibre-gl';

// Maplibre's map type contains some kind of cyclic dependency that causes an infinite loop in immers's change
// tracking. To save it off, we wrap it in a JS closure so that Redux just checks the function for changes, rather
// than recursively walking the whole object.
type MapGetter = () => MaplibreMap | undefined;

export interface DashboardMapState {
  maplibreMap: MapGetter;
  layers: any[];
  bounds: LngLatBounds | null;
  zoom: number | null;
}

const initialState: DashboardMapState = {
  maplibreMap: (() => {}) as MapGetter,
  layers: [],
  bounds: null,
  zoom: null,
};

export const dashboardMapStateSlice = createSlice({
  name: 'dashboardMapState',
  initialState,
  reducers: {
    setDashboardMap: (state, action: PayloadAction<MapGetter>) => ({
      ...state,
      maplibreMap: action.payload,
    }),
    addDashboardLayer: (state, action: PayloadAction<any>) => ({
      ...state,
      layers: [...state.layers, action.payload],
    }),
    removeDashboardLayer: (state, action: PayloadAction<string>) => ({
      ...state,
      layers: state.layers.filter(layer => layer.id !== action.payload),
    }),
    setDashboardLayers: (state, action: PayloadAction<any[]>) => ({
      ...state,
      layers: action.payload,
    }),
    setDashboardBounds: (state, action: PayloadAction<LngLatBounds>) => ({
      ...state,
      bounds: action.payload,
    }),
    setDashboardLocation: (
      state,
      action: PayloadAction<{ bounds: LngLatBounds; zoom: number }>,
    ) => ({
      ...state,
      bounds: action.payload.bounds,
      zoom: action.payload.zoom,
    }),
  },
});

// Getters
export const dashboardMapSelector = (state: {
  dashboardMapState: DashboardMapState;
}): MaplibreMap | undefined => state.dashboardMapState.maplibreMap();

export const dashboardLayersSelector = (state: {
  dashboardMapState: DashboardMapState;
}): any[] => state.dashboardMapState.layers;

export const dashboardBoundsSelector = (state: {
  dashboardMapState: DashboardMapState;
}): LngLatBounds | null => state.dashboardMapState.bounds;

export const dashboardZoomSelector = (state: {
  dashboardMapState: DashboardMapState;
}): number | null => state.dashboardMapState.zoom;

// Setters
export const {
  setDashboardMap,
  addDashboardLayer,
  removeDashboardLayer,
  setDashboardLayers,
  setDashboardBounds,
  setDashboardLocation,
} = dashboardMapStateSlice.actions;

export default dashboardMapStateSlice.reducer;
