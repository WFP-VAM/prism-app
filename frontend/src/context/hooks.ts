import { useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from 'context/store';

// See https://react-redux.js.org/tutorials/typescript-quick-start#define-typed-hooks
// Use these hooks throughout the app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
