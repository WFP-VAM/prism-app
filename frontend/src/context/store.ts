import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateReduce from './analysisResultStateSlice';
import mapSelectionLayerStateReduce from './mapSelectionLayerStateSlice';
import mapBoundaryInfoStateReduce from './mapBoundaryInfoStateSlice';
import datasetResultStateReduce from './datasetStateSlice';
import mapTileLoadingStateReduce from './mapTileLoadingStateSlice';
import leftPanelStateReduce from './leftPanelStateSlice';
import opacityStateReduce from './opacityStateSlice';
import anticipatoryActionDroughtStateReduce from './anticipatoryAction/AADroughtStateSlice';
import anticipatoryActionStormStateReduce from './anticipatoryAction/AAStormStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateReduce,
  notificationState: notificationStateReduce,
  mapSelectionLayerStateSlice: mapSelectionLayerStateReduce,
  mapBoundaryInfoState: mapBoundaryInfoStateReduce,
  datasetState: datasetResultStateReduce,
  mapTileLoadingState: mapTileLoadingStateReduce,
  leftPanelState: leftPanelStateReduce,
  opacityState: opacityStateReduce,
  anticipatoryActionDroughtState: anticipatoryActionDroughtStateReduce,
  anticipatoryActionStormState: anticipatoryActionStormStateReduce,
});

export const store = configureStore({
  reducer,
  middleware: getDefaultMiddleware({
    // TODO: Instead of snoozing this check, we might want to
    // serialize the state
    serializableCheck: false,
    immutableCheck: {
      // do not check the following state branches for accidental
      // mutations. This saves a lot of time in dev mode (but has no
      // impact on production builds).
      ignoredPaths: [
        'mapState.boundaryRelations',
        'mapState.layersData',
        'analysisResultState.result',
        'serverState.availableDates',
      ],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
