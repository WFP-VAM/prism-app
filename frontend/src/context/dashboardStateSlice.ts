import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { appConfig } from 'config';
import { DashboardElementType } from 'config/types';
import { Map as MaplibreMap } from 'maplibre-gl';
import { MapState, DateRange } from 'context/mapStateSlice';
import { LayerDefinitions } from 'config/utils';
import { keepLayer } from 'utils/keep-layer-utils';
import { BoundaryRelationsDict } from 'components/Common/BoundaryDropdown/utils';
import { getLayerMapId } from 'utils/map-utils';

import type {
  ConfiguredReport,
  LayerType,
  DashboardMapConfig,
} from 'config/types';
import type { RootState } from './store';

type MapGetter = () => MaplibreMap | undefined;

interface OpacityEntry {
  mapLayerId: string;
  opacityType: string;
  value: number;
}

interface SetDashboardOpacityParams {
  map: MaplibreMap | undefined;
  layerId: LayerType['id'] | undefined;
  layerType: LayerType['type'] | 'analysis' | undefined;
  value: number;
  callback?: (v: number) => void;
}

export interface DashboardMapState extends MapState {
  opacityMap: { [key: string]: OpacityEntry };
}

export type DashboardMode = 'edit' | 'preview';

export interface DashboardState {
  selectedDashboardIndex: number;
  title: string;
  mode: DashboardMode;
  flexElements: ConfiguredReport['flexElements'];
  maps: DashboardMapState[];
  syncMapsEnabled: boolean;
  sharedViewport?: {
    bounds: [number, number, number, number]; // [west, south, east, north]
    zoom: number;
  };
}

const getDashboardConfig = (index: number) => {
  if (
    !Array.isArray(appConfig.configuredReports) ||
    appConfig.configuredReports.length === 0
  ) {
    // Return a fallback config object or throw an error
    return { title: 'Dashboard', flexElements: [], maps: [] };
  }
  return appConfig.configuredReports[index] || appConfig.configuredReports[0];
};

const getMapLayerOpacityConfig = (
  layerId: LayerType['id'],
  layerType: LayerType['type'] | 'analysis',
): [string, string] => {
  switch (layerType) {
    case 'wms':
    case 'static_raster':
      return [getLayerMapId(layerId), 'raster-opacity'];
    case 'admin_level_data':
    case 'composite':
    case 'impact':
    case 'geojson_polygon':
      return [getLayerMapId(layerId), 'fill-opacity'];
    case 'point_data':
      if (layerId?.includes('_report')) {
        return [getLayerMapId(layerId), 'fill-opacity'];
      }
      return [getLayerMapId(layerId), 'circle-opacity'];
    case 'analysis':
      return ['layer-analysis', 'fill-opacity'];
    default:
      return [getLayerMapId(layerId), 'fill-opacity'];
  }
};

const createInitialState = (dashboardIndex: number = 0): DashboardState => {
  const dashboardConfig = getDashboardConfig(dashboardIndex);

  return {
    selectedDashboardIndex: dashboardIndex,
    title: dashboardConfig?.title || 'Dashboard',
    mode: 'preview' as DashboardMode,
    flexElements: dashboardConfig?.flexElements || [],
    syncMapsEnabled: false,
    sharedViewport: undefined,
    maps:
      dashboardConfig?.maps?.map((mapConfig: DashboardMapConfig) => {
        // Process pre-selected layers
        const preSelectedLayers: LayerType[] = [];
        const initialOpacityMap: { [key: string]: OpacityEntry } = {};

        if (mapConfig.preSelectedMapLayers) {
          mapConfig.preSelectedMapLayers.forEach(layerConfig => {
            const layerId =
              typeof layerConfig === 'string'
                ? layerConfig
                : layerConfig.layerId;
            const opacity =
              typeof layerConfig === 'string'
                ? 1.0
                : (layerConfig.opacity ?? 1.0);

            const layer = LayerDefinitions[layerId];
            if (layer) {
              // eslint-disable-next-line fp/no-mutating-methods
              preSelectedLayers.push(layer);
              const [mapLayerId, opacityType] = getMapLayerOpacityConfig(
                layerId,
                layer.type,
              );
              // eslint-disable-next-line fp/no-mutation
              initialOpacityMap[layerId] = {
                mapLayerId,
                opacityType,
                value: opacity,
              };
            } else {
              console.warn(
                `Pre-selected layer "${layerId}" not found in LayerDefinitions`,
              );
            }
          });
        }

        return {
          layers: preSelectedLayers,
          dateRange: {
            startDate: mapConfig.defaultDate
              ? new Date(mapConfig.defaultDate).getTime()
              : undefined,
          },
          maplibreMap: () => undefined,
          errors: [],
          layersData: [],
          loadingLayerIds: [],
          boundaryRelationData: {},
          opacityMap: initialOpacityMap,
          minMapBounds: mapConfig.minMapBounds,
        };
      }) || [],
  };
};

const initialState: DashboardState = createInitialState();

export const dashboardStateSlice = createSlice({
  name: 'dashboardState',
  initialState,
  reducers: {
    setSelectedDashboard: (_state, action: PayloadAction<number>) => {
      const dashboardIndex = action.payload;
      return createInitialState(dashboardIndex);
    },
    toggleMapSync: state => {
      const newSyncEnabled = !state.syncMapsEnabled;
      return {
        ...state,
        syncMapsEnabled: newSyncEnabled,
        sharedViewport: newSyncEnabled ? state.sharedViewport : undefined,
      };
    },
    setSharedViewport: (
      state,
      action: PayloadAction<{
        bounds: [number, number, number, number];
        zoom: number;
      }>,
    ) => ({
      ...state,
      sharedViewport: action.payload,
    }),
    setTitle: (state, action: PayloadAction<string>) => ({
      ...state,
      title: action.payload,
    }),
    setMode: (state, action: PayloadAction<DashboardMode>) => ({
      ...state,
      mode: action.payload,
    }),
    setTextContent: (
      state,
      action: PayloadAction<{ index: number; content: string }>,
    ) => {
      const { index, content } = action.payload;
      return {
        ...state,
        flexElements: state.flexElements.map((element, i) =>
          i === index ? { type: DashboardElementType.TEXT, content } : element,
        ),
      };
    },
    addLayerToMap: (
      state,
      action: PayloadAction<{ index: number; layer: LayerType }>,
    ) => {
      const { index, layer } = action.payload;
      const layersToAdd = layer?.group?.activateAll
        ? Object.values(LayerDefinitions).filter(l =>
            layer?.group?.layers?.map(subLayer => subLayer.id).includes(l.id),
          )
        : [layer];
      const filteredLayers = state.maps[index].layers.filter(l =>
        keepLayer(l, layer),
      );

      const newLayers =
        layer.type === 'boundary'
          ? [...filteredLayers, ...layersToAdd]
          : [...filteredLayers, ...layersToAdd];

      const dedupedLayers = newLayers.filter(
        (l, i, self) =>
          i === self.findIndex(t => t.id === l.id && t.type === l.type),
      );

      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? {
                ...mapInstance,
                layers: dedupedLayers,
              }
            : mapInstance,
        ),
      };
    },
    removeLayerFromMap: (
      state,
      action: PayloadAction<{ index: number; layer: LayerType }>,
    ) => {
      const { index, layer } = action.payload;
      const filteredLayers = state.maps[index].layers.filter(l =>
        keepLayer(l, layer),
      );

      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? { ...mapInstance, layers: filteredLayers }
            : mapInstance,
        ),
      };
    },
    updateMapDateRange: (
      state,
      action: PayloadAction<{ index: number; dateRange: DateRange }>,
    ) => {
      const { index, dateRange } = action.payload;

      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index ? { ...mapInstance, dateRange } : mapInstance,
        ),
      };
    },
    setMap: (
      state,
      action: PayloadAction<{ index: number; maplibreMap: MapGetter }>,
    ) => {
      const { index, maplibreMap } = action.payload;
      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index ? { ...mapInstance, maplibreMap } : mapInstance,
        ),
      };
    },
    removeLayerData: (
      state,
      action: PayloadAction<{ index: number; layer: LayerType }>,
    ) => {
      const { index, layer } = action.payload;
      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? {
                ...mapInstance,
                layersData: mapInstance.layersData.filter(
                  ({ layer: dataLayer }) => dataLayer.id !== layer.id,
                ),
              }
            : mapInstance,
        ),
      };
    },
    setBoundaryRelationData: (
      state,
      action: PayloadAction<{ index: number; data: BoundaryRelationsDict }>,
    ) => {
      const { index, data } = action.payload;
      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? { ...mapInstance, boundaryRelationData: data }
            : mapInstance,
        ),
      };
    },
    dismissError: (
      state,
      action: PayloadAction<{ index: number; error: string }>,
    ) => {
      const { index, error } = action.payload;
      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? {
                ...mapInstance,
                errors: mapInstance.errors.filter(msg => msg !== error),
              }
            : mapInstance,
        ),
      };
    },
    setDashboardOpacity: (
      state,
      action: PayloadAction<{ index: number } & SetDashboardOpacityParams>,
    ) => {
      const { index, map, layerId, layerType, value, callback } =
        action.payload;

      if (!map || !layerId || value === undefined || !layerType) {
        return state;
      }

      const [mapLayerId, opacityType] = getMapLayerOpacityConfig(
        layerId,
        layerType,
      );

      // update map
      if (map.getLayer(mapLayerId) !== undefined && value !== undefined) {
        map.setPaintProperty(mapLayerId, opacityType, value);
        // force a update of the map style to ensure the change is reflected
        // see https://github.com/maplibre/maplibre-gl-js/issues/3373
        // TODO - check if the above issue got resolved from time to time.
        // eslint-disable-next-line no-underscore-dangle
        map.style._updateLayer(mapLayerId as any);
      }

      if (callback !== undefined) {
        callback(value);
      }

      return {
        ...state,
        maps: state.maps.map((mapInstance, i) =>
          i === index
            ? {
                ...mapInstance,
                opacityMap: {
                  ...mapInstance.opacityMap,
                  [layerId]: {
                    mapLayerId,
                    opacityType,
                    value,
                  },
                },
              }
            : mapInstance,
        ),
      };
    },
  },
});

// Getters
export const selectedDashboardIndexSelector = (state: RootState): number =>
  state.dashboardState.selectedDashboardIndex;

export const dashboardTitleSelector = (state: RootState): string =>
  state.dashboardState.title;

export const dashboardModeSelector = (state: RootState): DashboardMode =>
  state.dashboardState.mode;

export const dashboardFlexElementsSelector = (
  state: RootState,
): ConfiguredReport['flexElements'] => state.dashboardState.flexElements;

export const dashboardMapsSelector = (state: RootState): DashboardMapState[] =>
  state.dashboardState.maps;

export const dashboardSyncEnabledSelector = (state: RootState): boolean =>
  state.dashboardState.syncMapsEnabled;

export const dashboardSharedViewportSelector = (
  state: RootState,
): DashboardState['sharedViewport'] => state.dashboardState.sharedViewport;

export const dashboardOpacitySelector =
  (index: number, layerId: string) =>
  (state: RootState): number | undefined =>
    state.dashboardState.maps[index]?.opacityMap[layerId]?.value;

// Setters
export const {
  setSelectedDashboard,
  toggleMapSync,
  setSharedViewport,
  addLayerToMap,
  removeLayerFromMap,
  setTitle,
  setMode,
  setTextContent,
  updateMapDateRange,
  setMap,
  removeLayerData,
  setBoundaryRelationData,
  dismissError,
  setDashboardOpacity,
} = dashboardStateSlice.actions;

export default dashboardStateSlice.reducer;
