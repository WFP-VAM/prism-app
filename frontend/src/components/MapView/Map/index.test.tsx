import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';
import { store } from 'context/store';
import MapComponent from '.';

jest.mock('react-mapbox-gl', () => ({
  __esModule: true,
  default: () => () => {
    return <span>Mock map</span>;
  },
  Cluster: () => <span>Mock cluster</span>,
}));

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    replace: jest.fn(),
    location: {
      search: '',
    },
  }),
}));

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <MapComponent setIsAlertFormOpen={() => {}} panelHidden={false} />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
