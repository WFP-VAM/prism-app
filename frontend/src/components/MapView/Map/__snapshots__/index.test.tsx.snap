// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    style="position: relative; width: 100%; height: 100%;"
  />
</div>
`;
